Get multi-sequences to work for v3 ranges/seqs

MImranAsghar opened the pull request:
- Update sequences and canvases selectors to get multi-sequences to work for v3 top-level ranges with behavior sequence
- Add integration test for v3 multi sequences

MImranAsghar committed:
Get multi-sequences to work for v3 ranges/seqs

- Update sequences and canvases selectors to get multi-sequences to work for v3 top-level ranges with behavior sequence, check if full canvases exist in each sequence
- Add integration test for v3 multi sequences

mejackreed commented:
Thanks for moving this forward @MImranAsghar . I'm seeing some issues here which seems related. Can you investigate? https://deploy-preview-3333--mirador-dev.netlify.app/__tests__/integration/mirador/toc.html

MImranAsghar commented:
Thanks Jack. Yes, looking into it.

MImranAsghar committed:
Add multi-seq select option to display..

... all items including ToC for v3, other changes
- Display an option in multi seq select bar that displays all the canvases in v3 sequence along with ToC
- Constrain the size of multi seqs select bar
- Change sequences test to expect uuid instead of undefined id for first v3 sequence
- Add checkes in getSequences selector before setting id for first v3 sequence which has undefined id initially

MImranAsghar commented:
Hi @mejackreed , i have added some more changes. I think for the tests `  ● how plugins relate to state › plugin can catch core action in its reducer
` is the one that is timing out. any suggestions of why this could be?

MImranAsghar commented:
when running these tests on mirador multiple times, some tests are giving timeout in a particular run but when I run the tests again the timeout disappears. Is it okay if those tests fail(timeout) in a pull request in that case?

mejackreed commented on src/state/selectors/sequences.js:
```diff
@@ -16,11 +17,35 @@ export const getSequences = createSelector(
+    // v3 sequence (not range sequence): assign id if not there
+    const manifestSequences = manifest.getSequences();
+    if (v3RangeSequences && manifestSequences && manifestSequences.length > 0
+       && !manifestSequences[0].id) {
```
Can you say more about the case where an id would not be present?

mejackreed reviewed (changes requested):
Thanks for continuing to move this forward @MImranAsghar . However, I'm seeing some issues with previous working content: `https://iiif.bodleian.ox.ac.uk/iiif/manifest/e32a277e-91e2-4a6d-8ba6-cc4bad230410.json` 

It would also be helpful I think to either (or both) add some additional unit tests in the modified selectors to better codify the changes that are happening. I'm still not fully clear on what the modeling changes are intended to represent.

MImranAsghar commented on src/state/selectors/canvases.js:
```diff
@@ -41,11 +49,24 @@ export const getCurrentCanvas = createSelector(
     if (!sequence || !window) return undefined;
 
-    if (!window.canvasId) return sequence.getCanvasByIndex(0);
+    if (!window.canvasId && (typeof sequence.getCanvasByIndex == 'function')) {
```
Hi Jack, this is for when the sequence represents a 'Range' manifesto object here which does not provide the 'getCanvasById' or 'getCanvases' function. 'Range' manifesto object has 'items' where it lists all the canvases.

MImranAsghar commented on src/containers/WindowSideBarCanvasPanel.js:
```diff
@@ -65,15 +65,24 @@ const styles = theme => ({
     paddingLeft: theme.spacing(1),
   },
   select: {
+    '& .MuiSelect-icon': {
```
Hi Jack, this is for constaining the size of the select dropdown so it does not extend beyond the sidebar panel when there is a long sequence label.. the reasoning for setting background color to white was to make sure that the 'Arrow icon' in the select bar and the label don't overlap.

MImranAsghar committed:
Update seq select icon background ...
color, update getCurrentCanvas

MImranAsghar commented on src/state/selectors/sequences.js:
```diff
@@ -16,11 +17,44 @@ export const getSequences = createSelector(
+      /**  Add manifesto canvases (items) to the ranges if 'items' property of ranges
+       *  is empty
+       */
+      if (v3RangeSequences.length > 0 && manifest.items && manifest.items.length > 0
+          && manifest.items[0].items && manifest.items[0].items.length > 0) {
+        /** Use manifesto canvases provided in manifest 'items' property to populate
+         * range 'items'/canvases
+        */
+        const canvases = manifest.items[0].items;
+
+        v3RangeSequences.map((sequence) => {
+          if (sequence.items.length === 0) {
+            const updatedSequence = sequence;
+            updatedSequence.items = sequence.canvases.map(canvasId => {
+              const fullCanvas = canvases.find(c => c.id === canvasId);
+              return fullCanvas;
+            });
+            return updatedSequence;
+          }
+          return sequence;
+        });
+      }
+    }
```
Hi @mejackreed , I made a pull request to 'get' canvases in the iiif manifesto library as well so this specific change won't be needed in future and i will remove it later after that pull request gets accepted. The manifesto pull request is here https://github.com/IIIF-Commons/manifesto/pull/79. Also I was wondering if you got a chance to get feedback from the designer on whether the additional sequence (which includes all manifest canvases) can be created and displayed (even if its not listed as seqs). I can make the changes accordingly after feedback from the designer.

MImranAsghar committed:
Update startCanvas sagas to get start

... canvas based on the sequence/range

marlo-longley commented:
Closing this PR as it is pretty stale now and depends on some outside work in Manifesto (that PR is tracked in the ticket #2949)
Leaving #2949 open as it tracks the underlying need. Can reference this work in a new PR.
